### PR TITLE
New quality metrics

### PR DESCRIPTION
**Pull request by johnmartinstw**
### Goal
Creating and exposing new quality metrics from `AVFoundationPlayback`. 

### How to test
- Run test suite.

### CODE REVIEW

File: Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                            name: NSNotification.Name.AVPlayerItemNewAccessLogEntry,
+                            object: player.currentItem
+                        )
+
+                        expect(avfoundationPlayback.droppedFrames).toEventually(equal(2))
+                    }
+                }
+            }
+
+            context("when call play") {
```
Comment (pauloSantosMobileTW) [inline Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: Make sense move te common code to the context and have only the "expect" inside each "it" ?

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift [hunk 2]
```diff
@@ -30,6 +31,15 @@ open class AVFoundationPlayback: Playback {
 
     private var observers = [NSKeyValueObservation]()
 
+    private var lastLogEvent: AVPlayerItemAccessLogEvent? { player?.currentItem?.accessLog()?.events.last }
+
+    open var bitrate: Double? { lastLogEvent?.indicatedBitrate }
+    open var bandwidth: Double? { lastLogEvent?.observedBitrate }
+    open var averageBitrate: Double? { lastLogEvent?.averageVideoBitrate }
+    open var droppedFrames: Int? { lastLogEvent?.numberOfDroppedVideoFrames }
+    open var decodedFrames: Int? { -1 }
```
[the remark this answers is not in the document]
Comment (paulogama14) [reply Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift]: Yes, we made a research about this value on `AVPlayer` and didn't get a safe and reliable way to expose it. There is no documentation on AVPlayer context and the other suggestions are expensives calcules and without confidence. Do you have any idea?

File: Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift (hunk 1 above)
Comment (paulogama14) [reply Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: I don't like this approach, in my opinion the test should have an explicit behavior. We follow the AAA pattern for tests and omitting code can brake this rule. What do you think?

File: Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                }
+
+                stubsDescriptor?.name = "StubToHighlineVideo.mp4"
+            }
+
+            afterEach {
+                OHTTPStubsHelper.removeStub(with: stubsDescriptor)
+            }
+
+            context("when receives AVPlayerItemNewAccessLogEntry event") {
```
[the remark this answers is not in the document]
Comment (paulogama14) [reply Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: What do you mean? The stub for contexts?

[the remark this answers is not in the document]
Comment (pauloSantosMobileTW) [reply Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: The assemble code, something like that:
`                      let options = [
                            kSourceUrl: "http://clappr.io/highline.mp4"
                        ]
                        let avfoundationPlayback = AVFoundationPlayback(options: options)
                        let accessLog = AccessLogEventMock()
                        let player = PlayerMock(accessLogEvent: accessLog)
                        avfoundationPlayback.player = player
                        avfoundationPlayback.addObservers()

                        it("changes the bitrate") {
                            accessLog.setIndicatedBitrate(7.0)
                            NotificationCenter.default.post(
                                name: NSNotification.Name.AVPlayerItemNewAccessLogEntry,
                                object: player.currentItem
                        )

                        expect(avfoundationPlayback.bitrate).toEventually(equal(7.0))
`

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (pauloSantosMobileTW) [reply Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift]: LGTM this aproach, I can not find another solution too for that, thanks for sharing.

File: Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift (hunk 1 above)
Comment (pauloSantosMobileTW) [reply Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: I think the context can be saw as part of the AAA pattern, I always imagine like I am reading a book when reading a test and the duplicity makes the read harder in this case to me.

File: Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift [hunk 4]
```diff
@@ -0,0 +1,117 @@
+import Quick
+import Nimble
+import OHHTTPStubs
+
+@testable import Clappr
+
+class AVFoundationPlaybackQualityMetricsTests: QuickSpec {
+    override func spec() {
+        describe(".AVFoundationPlaybackQualityMetricsTests") {
```
Comment (lfonseca) [inline Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: Are these tests supposed to be integration ones?

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift [hunk 5]
```diff
@@ -323,8 +321,8 @@ open class AVFoundationPlayback: Playback {
     @objc func onAccessLogEntry(notification: NSNotification?) {
         guard lastBitrate != bitrate else { return }
         lastBitrate = bitrate
-        if let newBitrate = lastBitrate, !newBitrate.isNaN {
-            trigger(.didUpdateBitrate, userInfo: ["bitrate": newBitrate])
+        if let lastBitrate = lastBitrate, !lastBitrate.isNaN {
```
Comment (lfonseca) [inline Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift]: 👏 👏

File: Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift (hunk 4 above)
Comment (paulogama14) [reply Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: No, we've changed the header from response and now the stub are working. Could you check, please?

Comment (lfonseca) [reply Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: So maybe you should call the `onAccessLogEntry` method directly and config some spy object to catch the event sent.

Comment (paulogama14) [reply Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackQualityMetricsTests.swift]: Nice =]
Now we are checking with a spy object the event sent, the other tests don't have a `kSourceURL`.

**Pull request merged**